cli/spec.py

```diff
@@ -0,0 +1,260 @@
+    :param config: the infrared configuration file
+    :param args: additional arguments to pass to the clg.
+    """
+
+    res_args, subcommand_opts = _get_command_line_args(args, config,
+                                                       module_name)
+
+    # get the default values
+    command_args = subcommand_opts[res_args['command0']]
+    defaults = {k: v.get('default', None)
+                for k, v in command_args.iteritems()}
+
+    # generate config file if required
+    if res_args.get('generate-conf-file', None):
+        try:
+            file_name = res_args['generate-conf-file']
+            out_config = ConfigParser.ConfigParser()
+
+            section = res_args['command0']
+            out_config.add_section(section)
+            for opt, value in defaults.iteritems():
+                if value is not None:
+                    out_config.set(section, opt, value)
+            with open(file_name, 'w') as configfile:  # save
+                out_config.write(configfile)
+        except Exception as ex:
+            raise exceptions.IRException(ex.message)
+
+    # override defaults with env variables
+    for arg_name, arg_value in res_args.iteritems():
+        upper_arg_name = arg_name.upper()
+        if arg_value is None and upper_arg_name in os.environ:
+            defaults[arg_name] = os.getenv(upper_arg_name)
+
+    # override defaults with the ini file args
+    if 'from-file' in res_args:
+        file_args = res_args['from-file']
+        if file_args is not None and res_args['command0'] in file_args:
+            defaults = utils.dict_merge(
+                file_args[res_args['command0']],
+                defaults,
+                conflict_resolver=utils.ConflictResolver.none_resolver)
+
+    # replace defaults with cli
+    utils.dict_merge(res_args, defaults,
+                     conflict_resolver=utils.ConflictResolver.none_resolver)
+
+    # check if we have all the required arguments defined at the end of merge
+    unset_args = []
+    only_args = []
+    for arg_name, arg_param in command_args.iteritems():
+        if 'required' in arg_param and \
+                arg_param['required'] and \
+                res_args.get(arg_name, None) is None:
+            unset_args.append(arg_name)
+        if 'requires_only' in arg_param:
+            only_args.extend(arg_param['requires_only'])
+    if len(only_args) > 0:
+        intersection = set(unset_args).intersection(set(only_args))
+        if len(intersection) > 0:
+            raise exceptions.IRConfigurationException(
+                "Required only input arguments {} are not set!".format(
+                    intersection))
+    elif len(unset_args) > 0:
+        raise exceptions.IRConfigurationException(
+            "Required input arguments {} are not set!".format(
+                unset_args))
+    return res_args
+
+
+def _get_command_line_args(args, config, module_name):
```

Comment:
Please consider rename this function to something more meaningful because it returns cmd line args ('res_args') and subcommand_opts also, or even think of split this function into two.